File: Python/day14/AOC.py
import re

def parse_input(input):
    reindeer_data = []
    for line in input:
        speed, fly_time, rest_time = map(int, re.findall(r"(\d+)", line))
        reindeer_data.append((speed, fly_time, rest_time))        
    return reindeer_data

def part1(input, race_time=1000):
    reideer_data = parse_input(input)
    times = []
    for reindeer in reideer_data:
        speed, fly_time, rest_time = reindeer
        time = 0
        distance = 0
        while time < race_time:
            if time % (fly_time + rest_time) < fly_time:
                distance += speed
            time += 1
        times.append(distance)
    return times

def part2(input, race_time=1000):
    reindeer_data = parse_input(input)
    points = [0] * len(reindeer_data)
    distances = [0]*len(reindeer_data)
    time = 0
    while time < race_time:
        for i, reindeer in enumerate(reindeer_data):
            speed, fly_time, rest_time = reindeer
            distance = distances[i]
            if time % (fly_time + rest_time) < fly_time:
                distance += speed
            distances[i] = distance
        for i, distance in enumerate(distances):
            if distance == max(distances):
                points[i] += 1
        time += 1
    return points

def main():
    input = open("input.txt", "r").read().splitlines()
    part1_sum = max(part1(input, 2503))
    print(f"Part 1: {part1_sum}")
    part2_sum = max(part2(input, 2503))
    print(f"Part 2: {part2_sum}")

if __name__ == "__main__":
    main()

File: Go/day14/AOC.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strings"
)

type Reindeer struct {
	name     string
	speed    int
	flyTime  int
	restTime int
}

func parseInput(input []string) []Reindeer {
	reindeer := []Reindeer{}

	var name string
	var speed, flyTime, restTime int
	for _, line := range input {
		fmt.Sscanf(line, "%s can fly %d km/s for %d seconds, but then must rest for %d seconds.", &name, &speed, &flyTime, &restTime)
		reindeer = append(reindeer, Reindeer{name, speed, flyTime, restTime})
	}

	return reindeer
}

func part1(input []string, racetime int) int {
	r := parseInput(input)
	var distance []int

	for _, deer := range r {
		cycleTime := deer.flyTime + deer.restTime
		cycles := racetime / cycleTime
		remainingTime := racetime % cycleTime
		distance = append(distance, (cycles*deer.flyTime+min(deer.flyTime, remainingTime))*deer.speed)
	}

	slices.Sort(distance)
	return distance[len(distance)-1]
}

func part2(input []string, racetime int) int {
	r := parseInput(input)
	points := make([]int, len(r))
	distance := make([]int, len(r))

	for i := 1; i <= racetime; i++ {
		furthest := 0
		for j, deer := range r {
			cycleTime := deer.flyTime + deer.restTime
			cycles := i / cycleTime
			remainingTime := i % cycleTime
			distance[j] = (cycles*deer.flyTime + min(deer.flyTime, remainingTime)) * deer.speed
			if distance[j] > furthest {
				furthest = distance[j]
			}
		}

		for j, d := range distance {
			if d == furthest {
				points[j]++
			}
		}
	}

	slices.Sort(points)
	return points[len(points)-1]
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines, 2503)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines, 2503)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
